firedrake/slate/slac/compiler.py

```diff
@@ -0,0 +1,324 @@
+"""This is SLATE's Linear Algebra Compiler. This module is responsible for generating
+C++ kernel functions representing symbolic linear algebra expressions written in SLATE.
+
+This linear algebra compiler uses both Firedrake's form compiler, the Two-Stage
+Form Compiler (TSFC) and COFFEE's kernel abstract syntax tree (AST) optimizer. TSFC
+provides this compiler with appropriate kernel functions (in C) for evaluating integral
+expressions (finite element variational forms written in UFL). COFFEE's AST optimizing
+framework produces the resulting kernel AST returned by: `compile_slate_expression`.
+
+The Eigen C++ library (http://eigen.tuxfamily.org/) is required, as all low-level numerical
+linear algebra operations are performed using the `Eigen::Matrix` methods built into Eigen.
+"""
+from __future__ import absolute_import, print_function, division
+
+import petsc
+
+from coffee import base as ast
+
+from firedrake.constant import Constant
+from firedrake.tsfc_interface import SplitKernel, KernelInfo
+from firedrake.slate.slate import (TensorBase, Tensor,
+                                   Transpose, Inverse, Negative,
+                                   Add, Sub, Mul, Action)
+from firedrake.slate.slac.kernel_builder import KernelBuilder
+from firedrake import op2
+
+from tsfc.parameters import SCALAR_TYPE
+
+from ufl.coefficient import Coefficient
+
+
+__all__ = ['compile_expression']
+
+
+def compile_expression(slate_expr, tsfc_parameters=None):
+    """Takes a SLATE expression `slate_expr` and returns the appropriate
+    :class:`firedrake.op2.Kernel` object representing the SLATE expression.
+
+    :arg slate_expr: a :class:'TensorBase' expression.
+
+    :arg tsfc_parameters: an optional `dict` of form compiler parameters to
+                          be passed onto TSFC during the compilation of ufl forms.
+    """
+    if not isinstance(slate_expr, TensorBase):
+        raise ValueError("Expecting a `slate.TensorBase` expression, not a %r" % slate_expr)
+
+    # TODO: Get PyOP2 to write into mixed dats
+    if any(len(a.function_space()) > 1 for a in slate_expr.arguments()):
+        raise NotImplementedError("Compiling mixed slate expressions")
+
+    # Initialize shape and statements list
+    shape = slate_expr.shape
+    statements = []
+
+    # Create a builder for the SLATE expression
+    builder = KernelBuilder(expression=slate_expr, tsfc_parameters=tsfc_parameters)
+
+    # Initialize coordinate and facet symbols
+    coordsym = ast.Symbol("coords")
+    coords = None
+    cellfacetsym = ast.Symbol("cell_facets")
+    inc = []
+
+    # Now we construct the list of statements to provide to the builder
+    context_temps = builder.temps.copy()
+    for exp, t in context_temps.items():
```

Comment:
I copy it because I don't want to actually alter the `dict` associated with these terminal temps. If I need to do something like `Action`, I will introduce a new temp for the coefficient in the `context_temps`. At the moment, I felt it was best to leave it separately rather than update the `temps` dict. I also use a different `Symbol` for this particular case (`Cn` I believe). This may be unnecessary though.